src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs

```diff
@@ -0,0 +1,183 @@
+                if (!root.UnsatisfiedPacks.Overlaps(branch.Packs))
+                {
+                    continue;
+                }
+
+                //construct the new condidate by combining the root and the branch
+                var combinedIds = new HashSet<WorkloadDefinitionId>(root.Workloads);
+                combinedIds.UnionWith(branch.Workloads);
+                var combinedPacks = new HashSet<WorkloadPackId>(root.Packs);
+                combinedPacks.UnionWith(branch.Packs);
+                var stillMissing = new HashSet<WorkloadPackId>(root.UnsatisfiedPacks);
+                stillMissing.ExceptWith(branch.Packs);
+                var candidate = new WorkloadSuggestionCandidate(combinedIds, combinedPacks, stillMissing);
+
+                //if the candidate contains all the requested packs, it's complete. else, recurse to try adding more branches to it.
+                if (candidate.IsComplete)
+                {
+                    completeSuggestions.Add(candidate);
+                }
+                else
+                {
+                    GatherCompleteSuggestions(candidate, branches, completeSuggestions);
+                }
+            }
+        }
+
+        public ICollection<WorkloadSuggestion> UnsortedSuggestions { get; }
+
+        /// <summary>
+        /// Finds the best value from a list of values according to one or more custom comparators. The comparators are an ordered fallback list - the second comparator is only checked when values are equal according to the first comparator, and so on.
+        /// </summary>
+        private T FindBest<T>(IEnumerable<T> values, params Comparison<T>[] comparators)
+        {
+            T best = values.First();
+
+            foreach (T val in values.Skip(1))
+            {
+                foreach (Comparison<T> c in comparators)
+                {
+                    var cmp = c(val, best);
+                    if (cmp > 0)
+                    {
+                        best = val;
+                        break;
+                    }
+                    else if (cmp < 0)
+                    {
+                        break;
+                    }
+                }
+            }
+            return best;
+        }
+
+        /// <summary>
+        /// Gets the suggestion with the lowest number of extra packs and lowest number of workload IDs.
+        /// </summary>
+        public WorkloadSuggestion GetBestSuggestion() => FindBest(
+                UnsortedSuggestions,
+                (x, y) => y.ExtraPacks - x.ExtraPacks,
+                (x, y) => y.Workloads.Count - x.Workloads.Count);
+
+        /// <summary>
+        /// A partial or complete suggestion for workloads to install, annotated with which requested packs it does not satisfy
+        /// </summary>
+        class WorkloadSuggestionCandidate : IEquatable<WorkloadSuggestionCandidate>
```

Comment:
nit: please add explicit modifier per editorconfig

Reply:
I have problem understanding what's this class present. It is a collection of Workloads , Packs but  GatherCompleteSuggestions has a list and a set of them (collection of collection of workload?). What does the branch mean here?